x/evm/keeper/keeper.go

```diff
@@ -0,0 +1,394 @@
+	// - storing account's Storage State
+	// - storing account's Code
+	// - storing transaction Logs
+	// - storing Bloom filters by block height. Needed for the Web3 API.
+	storeKey storetypes.StoreKey
+
+	// key to access the transient store, which is reset on every block during Commit
+	transientKey storetypes.StoreKey
+
+	// the address capable of executing a MsgUpdateParams message. Typically, this should be the x/gov module account.
+	authority sdk.AccAddress
+	// access to account state
+	accountKeeper types.AccountKeeper
+	// update balance and accounting operations with coins
+	bankKeeper types.BankKeeper
+	// access historical headers for EVM state transition execution
+	stakingKeeper types.StakingKeeper
+	// fetch EIP1559 base fee and parameters
+	feeMarketKeeper types.FeeMarketKeeper
+
+	// chain ID number obtained from the context's chain id
+	eip155ChainID *big.Int
+
+	// Tracer used to collect execution traces from the EVM transaction execution
+	tracer string
+
+	// EVM Hooks for tx post-processing
+	hooks types.EvmHooks
+
+	// custom stateless precompiled smart contracts
+	customPrecompiles evm.PrecompiledContracts
+
+	// evm constructor function
+	evmConstructor evm.Constructor
+	// Legacy subspace
+	ss paramstypes.Subspace
+}
+
+// NewKeeper generates new evm module keeper
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	storeKey, transientKey storetypes.StoreKey,
+	authority sdk.AccAddress,
+	ak types.AccountKeeper,
+	bankKeeper types.BankKeeper,
+	sk types.StakingKeeper,
+	fmk types.FeeMarketKeeper,
+	customPrecompiles evm.PrecompiledContracts,
+	evmConstructor evm.Constructor,
+	tracer string,
+	ss paramstypes.Subspace,
+) *Keeper {
+	// ensure evm module account is set
+	if addr := ak.GetModuleAddress(types.ModuleName); addr == nil {
+		panic("the EVM module account has not been set")
+	}
+
+	// ensure the authority account is correct
+	if err := sdk.VerifyAddressFormat(authority); err != nil {
+		panic(err)
+	}
+
+	// NOTE: we pass in the parameter space to the CommitStateDB in order to use custom denominations for the EVM operations
+	return &Keeper{
+		cdc:               cdc,
+		authority:         authority,
+		accountKeeper:     ak,
+		bankKeeper:        bankKeeper,
+		stakingKeeper:     sk,
+		feeMarketKeeper:   fmk,
+		storeKey:          storeKey,
+		transientKey:      transientKey,
+		customPrecompiles: customPrecompiles,
+		evmConstructor:    evmConstructor,
+		tracer:            tracer,
+		ss:                ss,
+	}
+}
+
+// Logger returns a module-specific logger.
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", types.ModuleName)
+}
+
+// WithChainID sets the chain id to the local variable in the keeper
+func (k *Keeper) WithChainID(ctx sdk.Context) {
+	chainID, err := ethermint.ParseChainID(ctx.ChainID())
+	if err != nil {
+		panic(err)
+	}
+
+	if k.eip155ChainID != nil && k.eip155ChainID.Cmp(chainID) != 0 {
+		panic("chain id already set")
```

Comment:
This panic had me baffelled once and I didn't find this message descriptive enough
```suggestion
		panic(fmt.Sprintf("chain id is already set and cannot be changed - trying to change from '%s' to '%s'", k.eip155ChainID, chainID))
```